src/libmongoc/src/mongoc/mongoc-client-pool.c

```diff
@@ -528,6 +529,15 @@ mongoc_client_pool_set_server_api (mongoc_client_pool_t *pool,
       return false;
    }
 
+   if (mongoc_client_pool_get_size (pool)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_POOL,
+                      MONGOC_ERROR_POOL_API_TOO_LATE,
+                      "Cannot set server api after a client has been created");
+      return false;
+   }
```

Comment:
To avoid the edge you mentioned of having the pool size fall back down to zero, alternatively, I think adding `bool api_set` to the pool, rather than the client (per Sam's comment), and setting it to `true` once a client has been created (popped) would solve the problem. Here we'd need to check that boolean (which would be immutable once changed), and not the pool size. Though, the name `api_set` might be confusing since the actual pointer to the API struct could still be `NULL`.

Reply:
Good point, I was considering that. I'll add a `client_created` flag that gets set automatically when the first client is created.